sklearn/metrics/pairwise.py

```diff
@@ -671,6 +686,260 @@ def cosine_distances(X, Y=None):
+
+    if not isinstance(scale, (bool, list, np.ndarray)):
+        raise TypeError("Parameter scale must be boolean, list, or ndarray")
+
+    if X is None or len(X) == 0:
+        raise ValueError("X can not be None or empty")
+
+    y_is_not_none = Y is not None
+
+    # It is necessary to convert to ndarray in advance to define the dtype
+    # as np.object, otherwise numeric columns will be converted to string
+    # if there are other string columns.
+    if not isinstance(X, np.ndarray):
+        X = np.asarray(X, dtype=np.object)
+
+    if y_is_not_none and not isinstance(Y, np.ndarray):
+        Y = np.asarray(Y, dtype=np.object)
+
+    X, Y = check_pairwise_arrays(X, Y, precomputed=False, dtype=X.dtype,
+                                 force_all_finite=False)
+
+    n_rows, n_cols = X.shape
+
+    categorical_features_obj, categorical_features_num, numerical_features = \
+        _detect_categorical_features(X, categorical_features)
+
+    # Categorical columns
+    X_cat_obj = X[:, categorical_features_obj].astype(np.object)
+    X_cat_num = X[:, categorical_features_num].astype(np.float)
+
+    # Numerical columns
+    X_num = X[:, numerical_features].astype(np.float)
+
+    Y_cat_obj = None
+    Y_cat_num = None
+    Y_num = None
+    if y_is_not_none:
+        Y_cat_obj = Y[:, categorical_features_obj].astype(np.object)
+        Y_cat_num = Y[:, categorical_features_num].astype(np.float)
+        Y_num = Y[:, numerical_features].astype(np.float)
+    else:
+        Y_cat_obj = X_cat_obj
+        Y_cat_num = X_cat_num
+        Y_num = X_num
+
+    num_cols = X_num.shape[1]
+    n_col_num_present = num_cols > 0
+    n_col_cat_obj_present = X_cat_obj.shape[1] > 0
+    n_col_cat_num_present = X_cat_num.shape[1] > 0
+
+    # Calculates the min and max values, and if requested, scale the
+    # input values in order to obtain the distances between 0 and 1,
+    # as proposed by the Gower's paper.
+    ranges = None
+    if n_col_num_present:
+        process_scale = False
+        if isinstance(scale, bool):
+            process_scale = scale
+        else:
+            if (isinstance(scale, list) and len(scale) != X_num.shape[1]) or\
+               (isinstance(scale, np.ndarray) and
+                    len(scale.flat) != X_num.shape[1]):
+                raise ValueError("Length of scale parameter must be equal "
+                                 "to the number of numerical columns.")
+            process_scale = True
+
+        ranges, min, max = _precompute_gower_params(X_num, Y_num, scale)
+
+        if not process_scale and (np.min(min) < 0 or np.max(max) > 1):
+            raise ValueError("Input data is not scaled between 0 and 1.")
+
+    y_n_rows = Y.shape[0]
+    D = np.zeros((n_rows, y_n_rows), dtype=np.float)
+
+    for i in range(n_rows):
+        j_start = i
+
+        # for non square results
+        if n_rows != y_n_rows or y_is_not_none:
+            j_start = 0
+        sum_s = 0
```

Comment:
Please add a heading/comment along the lines of:
```Compute D[i, j_start:]```

Alternatively, pull the guts of this out into a helper, like:

```
def gower_vec_mat(x, Y, ...):
```

Reply:
I think that you can make this code substantially shorter, particularly if we use NaN and not None for missing values in object columns (in accordance with pandas's default).

Something like:

```py
cat_dists = X[i, cat_mask] != Y[j_start:, cat_mask]
num_dists = abs(X[i, num_mask] - Y[j_start:, num_mask]) / ranges
n_missing = np.isnan(cat_dists).sum(axis=1) + np.isnan(num_dists).sum(axis=1)
D[i, j_start:] = np.nansum(cat_dists, axis=1) + np.nansum(num_dists, axis=1)) / n_missing
```
Please let me know if I'm missing something in my understanding.

Reply:
Some months ago I could not store None and nan in the same column of an numpy array, an error was thrown. Now I can see this is possible, something changed. 

I'll make some adjustments to give support for both None and nan, as both could be present in the same column, being Pandas or Numpy stuff. We can't force the user to replace None by nan in their data.

Reply:
>> Please add a heading/comment along the lines of:
>> Compute D[i, j_start:]
>> Alternatively, pull the guts of this out into a helper ....

Hi @jnothman , do you mean to comment and push the content of the big loop `for i in range(X.shape[0]):` into a helper function?

Reply:
>> Please let me know if I'm missing something in my understanding.
Yes, your proposal works very well!

Reply:
I'll need to replace `X[i, cat_mask] != Y[j_start:, cat_mask]` by `~np.equals(X[i, cat_mask],  Y[j_start:, cat_mask])` because != is returning different values in the presence of np.nan in other deployments like open blas and ubuntu_atlas. The unit tests are failing because of this.